FlowCrypt/Functionality/Mail Provider/Message Provider/Gmail+Message.swift

```diff
@@ -8,34 +8,83 @@
-            query.format = kGTLRGmailFormatRaw
+            Task {
+                let messageSize = try await self.fetchMessageSize(identifier: identifier)
+                let fetcher = self.createMessageFetcher(identifier: identifier)
+                fetcher.receivedProgressBlock = { _, received in
+                    let progress = min(Float(received)/messageSize, 1)
+                    progressHandler?(.download(progress))
+                }
+                fetcher.beginFetch { data, error in
+                    if let error = error {
+                        reject(GmailServiceError.providerError(error))
+                        return
+                    }
+
+                    guard let data = data,
+                          let dictionary = try? JSONSerialization.jsonObject(with: data, options: []) as? [String: Any],
+                          let raw = dictionary["raw"] as? String
+                    else {
+                        return reject(GmailServiceError.missedMessageInfo("raw"))
+                    }
 
+                    guard let data = GTLRDecodeWebSafeBase64(raw) else {
+                        return reject(GmailServiceError.missedMessageInfo("data"))
+                    }
+                    resolve(data)
+                }
+            }
+        }
+    }
+
+    private func fetchMessageSize(identifier: String) async throws -> Float {
+        try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<Float, Error>) in
+            let query = createMessageQuery(identifier: identifier, format: kGTLRGmailFormatMetadata)
             self.gmailService.executeQuery(query) { _, data, error in
                 if let error = error {
-                    reject(GmailServiceError.providerError(error))
+                    continuation.resume(throwing: GmailServiceError.providerError(error))
                     return
                 }
+
                 guard let gmailMessage = data as? GTLRGmail_Message else {
-                    return reject(AppErr.cast("GTLRGmail_Message"))
-                }
-                guard let raw = gmailMessage.raw else {
-                    return reject(GmailServiceError.missedMessageInfo("raw"))
+                    continuation.resume(throwing: AppErr.cast("GTLRGmail_Message"))
+                    return
                 }
 
-                guard let data = GTLRDecodeWebSafeBase64(raw) else {
-                    return reject(GmailServiceError.missedMessageInfo("data"))
+                guard let sizeEstimate = gmailMessage.sizeEstimate?.floatValue else {
+                    continuation.resume(throwing: GmailServiceError.missedMessageInfo("sizeEstimate"))
+                    return
                 }
-                resolve(data)
+
+                // google returns smaller estimated size
+                let totalSize = sizeEstimate * Float(1.3)
```

Comment:
I did the exact same fix on browser extension. Did you get this from there? :-)

Reply:
Haha, no, I haven't seen your fix, just calculated difference between real and estimated sizes 🙂 Usually it's around 30%

Reply:
Careful to cap it at 100% when calculating. Else sometimes user will see 103%..

I think the difference is in base64 encoding which tends to add about a third.

Reply:
Yes, I handle this case with setting `1` as max progress value:
```
let progress = min(Float(received)/messageSize, 1)
```